samples/bluetooth/direct_test_mode/src/fem/nrf21540.c

```diff
@@ -0,0 +1,526 @@
+#define NRF21540_SPI_COMMAND_WRITE 0x03
+
+#define NRF21540_BITS_CONFREG0_TX_GAIN_Pos 2
+#define NRF21540_BITS_CONFREG0_TX_GAIN_Msk \
+	(0x1F << NRF21540_BITS_CONFREG0_TX_GAIN_Pos)
+
+#define NRF21540_SPI_WRITE(_reg) \
+	((NRF21540_SPI_COMMAND_WRITE << NRF21540_SPI_COMMAND_Pos) | (_reg))
+
+#define NRF21540_SPI_READ(_reg) \
+	((NRF21540_SPI_COMMAND_READ << NRF21540_SPI_COMMAND_Pos) | (_reg))
+
+#define TX_EN_REG DT_REG_ADDR(DT_PHANDLE(DT_NODELABEL(nrf_radio_fem), \
+					 tx_en_gpios))
+#define TX_EN_PIN DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem), tx_en_gpios)
+#define TX_EN_FLAGS DT_GPIO_FLAGS(DT_NODELABEL(nrf_radio_fem), tx_en_gpios)
+
+#define RX_EN_REG DT_REG_ADDR(DT_PHANDLE(DT_NODELABEL(nrf_radio_fem), \
+					 rx_en_gpios))
+#define RX_EN_PIN DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem), rx_en_gpios)
+#define RX_EN_FLAGS DT_GPIO_FLAGS(DT_NODELABEL(nrf_radio_fem), rx_en_gpios)
+
+#define PDN_PORT DT_GPIO_LABEL(DT_NODELABEL(nrf_radio_fem), pdn_gpios)
+#define PDN_PIN DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem), pdn_gpios)
+#define PDN_FLAGS DT_GPIO_FLAGS(DT_NODELABEL(nrf_radio_fem), pdn_gpios)
+
+#define ANT_SEL_PORT DT_GPIO_LABEL(DT_NODELABEL(nrf_radio_fem), ant_sel_gpios)
+#define ANT_SEL_PIN DT_GPIO_PIN(DT_NODELABEL(nrf_radio_fem), ant_sel_gpios)
+#define ANT_SEL_FLAGS DT_GPIO_FLAGS(DT_NODELABEL(nrf_radio_fem), ant_sel_gpios)
+
+#define NRF21540_SPI DT_ALIAS(nrf21540_spi)
+#define NRF21540_SPI_LABEL DT_LABEL(NRF21540_SPI)
+#define CS_GPIO_PORT DT_GPIO_LABEL(NRF21540_SPI, cs_gpios)
+
+#define T_NCS_SCLK 1
+
+#define NRF21540_GPIO_POLARITY_GET(_dt_property)                     \
+	((DT_GPIO_FLAGS(DT_NODELABEL(nrf_radio_fem), _dt_property) & \
+	GPIO_ACTIVE_LOW) ? false : true)
+
+enum nrf21540_trx {
+	NRF21540_TX,
+	NRF21540_RX
+};
+
+struct gpio_data {
+	const struct device *port;
+	gpio_pin_t pin;
+};
+
+struct gpiote_pin {
+	uint32_t abs_pin;
+	uint8_t gpiote_channel;
+	bool active_high;
+};
+
+static struct nrf21540 {
+	const struct device *spi_dev;
+	struct spi_cs_control spi_cs;
+	struct spi_config spi_cfg;
+	struct gpiote_pin tx_en;
+	struct gpiote_pin rx_en;
+	struct gpio_data pdn;
+	struct gpio_data ant_sel;
+	gppi_channel_t timer_ch;
+	gppi_channel_t pin_set_ch;
+	gppi_channel_t pin_clr_ch;
+	nrfx_timer_t timer;
+	bool busy;
+} nrf21540_cfg = {
+	.spi_cs = {
+		.gpio_pin = DT_GPIO_PIN(NRF21540_SPI, cs_gpios),
+		.gpio_dt_flags = DT_GPIO_FLAGS(NRF21540_SPI, cs_gpios),
+		.delay = T_NCS_SCLK
+	},
+	.spi_cfg = {
+		.frequency = DT_PROP(NRF21540_SPI, clock_frequency),
+		.operation = (SPI_OP_MODE_MASTER | SPI_WORD_SET(8) |
+			      SPI_TRANSFER_MSB | SPI_LINES_SINGLE),
+		.slave = 0,
+	},
+	.timer = NRFX_TIMER_INSTANCE(NRF21540_TIMER_INSTANCE)
+};
+
+enum nrf21540_reg {
+	NRF21540_REG_CONFREG0 = 0x00,
+} nrf21540_reg;
+
+static uint32_t active_evt;
+static uint32_t deactive_evt;
+
+static inline nrf_radio_task_t nrf21540_radio_task_get(enum nrf21540_trx dir)
+{
+	return dir == NRF21540_TX ? NRF_RADIO_TASK_TXEN : NRF_RADIO_TASK_RXEN;
```

Comment:
Consider adding parentheses for clarity

```suggestion
	return (dir == NRF21540_TX) ? NRF_RADIO_TASK_TXEN : NRF_RADIO_TASK_RXEN;
```

Reply:
Done